Replace phpversion() calls with PHP_VERSION constant

Ayesh opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/55680#ticket

We use several `phpversion()` function calls in WordPress core, which I propose to be replaced with `PHP_VERSION` constants.

`phpversion()` return value and `PHP_VERSION` constant value are identical, but the former is about 3 times slower because it is a function call compared to a direct constant value lookup.

Simple benchmark:

```php

$start = microtime(true);
for ($i = 0; $i < 1000000; $i++) {phpversion();}
echo microtime(true) - $start;

echo "\r\n";

$start = microtime(true);
for ($i = 0; $i < 1000000; $i++) {PHP_VERSION;}
echo microtime(true) - $start;

```
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Ayesh committed:
Replace phpversion() calls with PHP_VERSION constant

mukeshpanchal27 commented on src/wp-cron.php:
```diff
@@ -19,7 +19,7 @@
 
 /* Don't make the request block till we finish, if possible. */
-if ( function_exists( 'fastcgi_finish_request' ) && version_compare( phpversion(), '7.0.16', '>=' ) ) {
+if ( function_exists( 'fastcgi_finish_request' ) && version_compare( PHP_VERSION, '7.0.16', '>=' ) ) {
```
@Ayesh Can we update version_compare to below

`version_compare( PHP_VERSION, '7.0.16', '>=' )`
Update to
`PHP_VERSION_ID >= 70016`

Ayesh replied:
Nice catch, thank you. Changed. I also took liberty in changing the condition order, so the version comparison is done prior to the `function_exits` function call.

jrfnl commented on tests/phpunit/tests/load/wpIsIniValueChangeable.php:
```diff
@@ -40,7 +40,7 @@ public function data_wp_is_ini_value_changeable() {
 		);
 
-		if ( extension_loaded( 'Tidy' ) && version_compare( PHP_VERSION, '7.0.0', '>' ) ) {
+		if ( extension_loaded( 'Tidy' ) && PHP_VERSION_ID > 70000 ) {
```
Is this comparison correct ? Or should it be `>=` ? Or maybe the better question is: why is the PHP 7.0 comparison done in the first place ? The `tidy.clean_output` ini directive was added in PHP 5.0.0 as far as I can see.

Note: I haven't been able to determine whether the "changability" was changed in PHP 7.0. Could be that that's the reason ?

P.S.: 💯 on the principle of this PR!

Ayesh replied:
Thanks a lot. I thought to only switch the `phpversion()`/`version_compare` to `PHP_VERSION_ID` and keep it `>` for the parity with existing code.

I don't see anything that changed in the tidy extension between [7.0.0...7.0.1](https://github.com/php/php-src/compare/php-7.0.1...php-7.0.0), and like you said, it's questionable why the PHP 7.0 comparison is done there.

Do you think we should go ahead and change it to `>=`, or perhaps bring it in a separate ticket to remove the version comparison altogether?

jrfnl replied:
Might be worth doing a `blame` first to trace back why it was added in the first place and to check that if there was a valid reason, it was added correctly.

Ayesh replied:
The oldest commit I could find out was [`c68f1ec772b3`](https://github.com/WordPress/wordpress-develop/commit/c68f1ec772b33f3e5d563936ba4002644c6d5207`), that actually attributes you @jrfnl and @ocean90, but the commit message does not indicate if the `> 7.0.0` was deliberate. The [ticket](https://core.trac.wordpress.org/ticket/32075) links to PHP bug [#44936](https://bugs.php.net/bug.php?id=44936) that reports the ini access value being different in minor versions, so my uninformed gut feeling is to not change the current `>` to `>=`, even though PHP 7.0.0 is too old for us to bother that much.

jrfnl replied:
😁 Oh cricky... in that case, there's bound to be a reason for it, but as it was from before git was used, my original commit messages which would explain the reason have been lost to time...

So, yeah, let's leave it.

hellofromtonya replied:
For consistency and micro-optimization, should the conditional check order be flipped?

SergeyBiryukov replied:
Flipped in 3b1a418.

costdev commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -697,7 +697,7 @@ public static function debug_data() {
 
 		if ( function_exists( 'php_sapi_name' ) ) {
-			$php_sapi = php_sapi_name();
+			$php_sapi = PHP_SAPI;
```
This change isn't the same as the rest in this PR or as the ticket describes.

Ayesh replied:
Thanks for this good catch, I was planning to make another ticket/PR with this change, but had this one spilled over. Reverted now.

costdev reviewed (changes requested):
Noted one unrelated change.

Also, are the `function_exists( 'phpversion' )` conditions still necessary? If so, would `defined( 'PHP_VERSION' )` be more appropriate with this change?

Ayesh commented:
@costdev you are right, it absolutely makes sense to change the `function_exists('phpversion`)` do the `defined`, which I have now done. 

Apart from this PR, I think we could use a good cleanup of these `function_exists` calls we have accumulated over the years, which are now either bundled into core, is past the version test. Makes the code cleaner, and a happy path for OPCache.

jrfnl commented:
> @costdev you are right, it absolutely makes sense to change the `function_exists('phpversion')` do the `defined`, which I have now done.

Actually.... I think that check can probably be removed...

If I remember correctly, those kind of checks will have been put in place for silly hosts which think adding `phpversion` to the [`disabled_functions`](https://www.php.net/manual/en/ini.core.php#ini.disable-functions) ini directive increases security (it does not).

I'm trying to think, but I can't remember any way (for hosts) to prevent PHP from declaring the `PHP_VERSION` constant.

Unless someone else comes up with a way that could be done, that `function_exists()`/`defined()` check is no longer needed.

> Apart from this PR, I think we could use a good cleanup of these `function_exists` calls we have accumulated over the years, which are now either bundled into core, is past the version test. Makes the code cleaner, and a happy path for OPCache.

➕ Agreed. I remember doing such a check for WP 5.3 (after the PHP 5.2 version drop), but I doubt it's been done since.

Ayesh committed:
Remove unnecessary PHP_VERSION availability check in WP_Debug_Data

This was changed in a previous commit to replace `phpversion()` with
PHP_VERSION constant check. Because it is not possible for this constant
to be disabled, this check is no longer necessary.

Ayesh commented:
> Unless someone else comes up with a way that could be done, that function_exists()/defined() check is no longer needed.

Happily got rid of that check, thank you!

jrfnl reviewed (approved):
:shipit:

SergeyBiryukov commented on src/wp-cron.php:
```diff
@@ -19,7 +19,7 @@
 
 /* Don't make the request block till we finish, if possible. */
-if ( function_exists( 'fastcgi_finish_request' ) && version_compare( phpversion(), '7.0.16', '>=' ) ) {
+if ( PHP_VERSION_ID >= 70016 && function_exists( 'fastcgi_finish_request' ) ) {
```
Is there a reason for reversing the order of the conditions here?

I think `function_exists( 'fastcgi_finish_request' )` is the "primary" condition here and should probably remain first, as the second one can be removed once the minimum supported PHP version is bumped to 7.1+.

jrfnl replied:
Micro-optimization:
A straight numeric comparison between two constant values is faster than a function call and if the comparison yields `false`, the function call won't even be executed anymore.

SergeyBiryukov replied:
Thanks! To me, that makes the condition slightly less readable, as the main check is now placed behind the supplemental one.

So I guess I'm not a fan of this specific micro-optimization, but I see the point and don't have any strong objections if that's preferred here 🙂

SergeyBiryukov committed:
Flip the condition for micro-optimization

SergeyBiryukov commented:
> `phpversion()` return value and `PHP_VERSION` constant value are identical, but the former is about 3 times slower because it is a function call compared to a direct constant value lookup.
> 
> Simple benchmark:
> 
> ```
> $start = microtime(true);
> for ($i = 0; $i < 1000000; $i++) {phpversion();}
> echo microtime(true) - $start;
> 
> echo "\r\n";
> 
> $start = microtime(true);
> for ($i = 0; $i < 1000000; $i++) {PHP_VERSION;}
> echo microtime(true) - $start;
> ```

I was curious about the results of this benchmark. On my install, after a few runs, `PHP_VERSION` on average appears to be 5 times faster than `phpversion()`, making the performance improvement even more noticeable:
```
phpversion(): 0.0367081165314
PHP_VERSION:  0.00715589523315
```

Should be good to go.

SergeyBiryukov commented:
> On my install, after a few runs, `PHP_VERSION` on overage appears to be 5 times faster than `phpversion()`, making the performance improvement even more noticeable

For reference, this was on PHP 8.0 and 8.1. On PHP 7.4, the difference is indeed closer to 3 times.

SergeyBiryukov commented:
Thanks for the PR! Merged in https://core.trac.wordpress.org/changeset/53426.

Ayesh commented:
Thank you so much!
